torch/csrc/jit/argument_spec.h

```diff
@@ -0,0 +1,180 @@
+#include <Python.h>
+#include <iostream>
+#include <vector>
+#include "torch/csrc/autograd/variable.h"
+#include "torch/csrc/utils/hash.h"
+
+namespace torch { namespace jit {
+
+// GraphExecutor creates specializations of Graphs for different dimensionalitities
+// and types of inputs.
+
+// ArgumentSpec represents one particular specialization.
+// It is designed so that it can be created, hashed, and compared quickly
+// since it is used along the hot-path of the JIT to check if the code
+// we have created is valid for the given inputs.
+
+// TensorInfoPOD is only used internally in ArgumentSpec
+// API users should use TensorInfo
+struct TensorInfoPOD {
+  // total size is 64-bit
+  unsigned int type : 8;
+  unsigned int defined : 1;
+  unsigned int requires_grad : 1;
+  int device : 22;
+  uint32_t total_dims; // all TensorInfoPODs are in ArgumentSpec's tensor_info() array.
+                       // total_dims is the total number of dimensions seen so far
+                       // in all previous members of tensor_info(), including this tensor
+                       // 2*total_dims becomes the offset into the sizes_strides list
+                       // for the _next_ tensor in the tensor_info array
+                       // for tensor 0, the offset is always 0
+};
+static_assert(sizeof(TensorInfoPOD) == sizeof(int64_t),
+  "TensorInfoPOD must be 64-bit struct for ArgumentSpec encoding to work");
+
+struct TensorInfo;
+
+struct ArgumentSpec {
+  ArgumentSpec(bool nograd, ArrayRef<autograd::Variable> tensors)
+  :  hash_code(0), ntensors(tensors.size()) {
+    // the first allocation creates enough data for 5-dim tensors
+    data.reserve((1 + 5*2)*ntensors);
+    // insert 0-filled data for the TensorInfoPOD structs
+    data.insert(data.begin(), ntensors, 0);
+    // and reinterpret our data array as these structs
+    TensorInfoPOD * pods = reinterpret_cast<TensorInfoPOD*>(data.data());
```

Comment:
oh god. Good we compile with `-fno-strict-aliasing`. This is actually important, because you're going to be writing to this memory, which might not invalidate some `uint32_t` pointers/cached values in the compiler.

Reply:
Heh, yeah -- I will try making data into a union type so this is kosher.